aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java

```diff
@@ -383,44 +383,76 @@ public synchronized void initialize(
+                final SqlCommand sqlCommand = sqlCommandFactory.deleteFor(modelSchema, item, condition);
+                if (sqlCommand == null || sqlCommand.sqlStatement() == null
+                        || !sqlCommand.hasCompiledSqlStatement()) {
                     itemDeleteListener.onError(new DataStoreException(
                             "No delete statement found for the Model: " +
                                     modelSchema.getName(),
                             AmplifyException.TODO_RECOVERY_SUGGESTION
                     ));
                 }
 
-                databaseConnectionHandle.beginTransaction();
-                databaseConnectionHandle.execSQL(sqlCommand.sqlStatement());
-                databaseConnectionHandle.setTransactionSuccessful();
-                databaseConnectionHandle.endTransaction();
-
-                final StorageItemChange.Record record = StorageItemChange.<T>builder()
-                        .item(item)
-                        .itemClass((Class<T>) item.getClass())
-                        .type(StorageItemChange.Type.DELETE)
-                        .initiator(initiator)
-                        .build()
-                        .toRecord(storageItemChangeConverter);
-                itemChangeSubject.onNext(record);
-                itemDeleteListener.onResult(record);
+                synchronized (sqlCommand.getCompiledSqlStatement()) {
+                    final SQLiteStatement compiledSqlStatement = sqlCommand.getCompiledSqlStatement();
+                    compiledSqlStatement.clearBindings();
+                    bindPreCompiledStatementWithFieldValues(compiledSqlStatement, sqlCommand.getSelectionArgs());
+                    // executeUpdateDelete returns the number of rows affected.
+                    deleteSuccess = compiledSqlStatement.executeUpdateDelete() > 0;
+                    compiledSqlStatement.clearBindings();
+                }
+
+                if (deleteSuccess) {
+                    // Do not publish item change if write did not go through.
+                    final StorageItemChange.Record record = StorageItemChange.<T>builder()
+                            .item(item)
+                            .itemClass((Class<T>) item.getClass())
+                            .type(StorageItemChange.Type.DELETE)
+                            .predicate(predicate)
+                            .initiator(initiator)
+                            .build()
+                            .toRecord(storageItemChangeConverter);
+                    itemChangeSubject.onNext(record);
+                    itemDeleteListener.onResult(record);
+                } else {
+                    //TODO: implement ConditionCheckFailedException to be cause of this error
+                    itemDeleteListener.onError(new DataStoreException("Predicate condition not met.",
+                            AmplifyException.TODO_RECOVERY_SUGGESTION));
+                }
```

Comment:
I'm not sure if it is appropriate for us to fire `onError(...)`, here. It is an expected outcome for a predicate not to match, and so it is a normal operation of the system for a delete to not occur, in this case.

Perhaps what we need is a way to dispatch different types of success results. For example, maybe we need:

```
interface LocalStorageAdapter {
    ...
    public <T extends Model> void delete(
            @NonNull T item,
            @NonNull StorageItemChange.Initiator initiator,
            @Nullable QueryPredicate predicate,
            @NonNull ResultListener<Optional<StorageItemChange.Record>> itemDeleteListener);
    ...
}
```
So that:
```
new ResultListener(/* on result */ result -> {
    if (result.isPresent()) {
        // deletion went through.
    } else {
        // predicate did not match.
    }
}, /* on error */ error -> {
    // we tried to delete the thing, after a predicate match, but that failed!
});
```

Or, alternately, just the existing contract:
```
interface LocalStorageAdapter {
    ...
    public <T extends Model> void delete(
            @NonNull T item,
            @NonNull StorageItemChange.Initiator initiator,
            @Nullable QueryPredicate predicate,
            @NonNull ResultListener<StorageItemChange.Record> itemDeleteListener);
    ...
}
```
But we find a way to communicate a "no-op" in the existing argument. Like maybe we set the `entry` field as `null`, and provide a method to check it?

```
new ResultListener(changeRecord -> {
    if (changeRecord.hasEntry()) {
        // Change went through ...
    } else {
        // Predicate did not match.
    }
}, error -> {
    // Failure to delete, even though we tried
});
```

Reply:
I initially had it implemented as a no-op, but I discussed with Francisco and he told me that condition check failure is a valid error thrown in this case. 

DynamoDB's "conditional write" also appears to have similar mechanism: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html

Reply:
Okay, cool! Yes, I've receives those conditional checks as 400 series from Dynamo before. Good to be consistent with other AWS systems.